### ISSUE TITLE
debian 11 (bullseye) pgadmin4-server package: ImportError: PyO3 modules may only be initialized once per interpreter process

### ISSUE THREAD

**Issue by olivluca** [pgadmin-org/pgadmin4#6718]
Versions 7.5 and 7.6 of the pgadmin4-server package for debian 11 don't work. The traceback in the log is

```
Traceback (most recent call last):
  File "./pgAdmin4.py", line 43, in <module>
    from eventlet import hubs
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/__init__.py", line 17, in <module>
    from eventlet import convenience
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/convenience.py", line 7, in <module>
    from eventlet.green import socket
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/green/socket.py", line 21, in <module>
    from eventlet.support import greendns
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/support/greendns.py", line 79, in <module>
    setattr(dns, pkg, import_patched('dns.' + pkg))
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/support/greendns.py", line 61, in import_patched
    return patcher.import_patched(module_name, **modules)
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/patcher.py", line 132, in import_patched
    return inject(
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/patcher.py", line 109, in inject
    module = __import__(module_name, {}, {}, module_name.split('.')[:-1])
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/dns/dnssecalgs/__init__.py", line 5, in <module>
    from dns.dnssecalgs.dsa import PrivateDSA, PrivateDSANSEC3SHA1
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/dns/dnssecalgs/dsa.py", line 4, in <module>
    from cryptography.hazmat.primitives import hashes
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/cryptography/hazmat/primitives/hashes.py", line 10, in <module>
    from cryptography.hazmat.bindings._rust import openssl as rust_openssl
ImportError: PyO3 modules may only be initialized once per interpreter process

```

Version 7.4 of the package works.

```
$ LC_ALL=C apt-cache policy pgadmin4-server
pgadmin4-server:
  Installed: 7.4
  Candidate: 7.6
  Version table:
     7.6 500
        500 https://ftp.postgresql.org/pub/pgadmin/pgadmin4/apt/bullseye pgadmin4/main amd64 Packages
     7.5 500
        500 https://ftp.postgresql.org/pub/pgadmin/pgadmin4/apt/bullseye pgadmin4/main amd64 Packages
 *** 7.4 500
        500 https://ftp.postgresql.org/pub/pgadmin/pgadmin4/apt/bullseye pgadmin4/main amd64 Packages
        100 /var/lib/dpkg/status
     7.3 500
        500 https://ftp.postgresql.org/pub/pgadmin/pgadmin4/apt/bullseye pgadmin4/main amd64 Packages
     7.2 500
        500 https://ftp.postgresql.org/pub/pgadmin/pgadmin4/apt/bullseye pgadmin4/main amd64 Packages
     7.1 500
        500 https://ftp.postgresql.org/pub/pgadmin/pgadmin4/apt/bullseye pgadmin4/main amd64 Packages
     7.0 500
        500 https://ftp.postgresql.org/pub/pgadmin/pgadmin4/apt/bullseye pgadmin4/main amd64 Packages
     6.21 500
        500 https://ftp.postgresql.org/pub/pgadmin/pgadmin4/apt/bullseye pgadmin4/main amd64 Packages

```

**Comment by nikhil-mohite** [comment]
@FaharAbbasRizvi ,
Can you please verify it once?

**Comment by olivluca** [comment]
In case it's relevant (it probably is) it's the only package I installed (neither pgadmin4 nor pgadmin4-desktop) and I'm running under uwsgi with this ini file:

```
[uwsgi]
processes = 1
threads = 25
chdir = /usr/pgadmin4/web/
wsgi-file = /usr/pgadmin4/web/pgAdmin4.wsgi
uid = pgadmin4
plugins = python3
mount=/pgadmin=pgAdmin4:app
virtualenv=/usr/pgadmin4/venv
manage-script-name = true
```

**Comment by FaharAbbasRizvi** [comment]
I am able to reproduce the issue on Debian 11

<img width="1010" alt="Screen Shot 2023-08-28 at 5 20 52 PM" src="https://github.com/pgadmin-org/pgadmin4/assets/43006122/4a537b1b-03cb-4e7e-a8ea-8ecc619f3ff5">


Could you please share the exact step to reproduce?

**Comment by FlossyWeb** [comment]
I've got the same issue here, everything's fine using pgadmin4-server apt package version 7.4 on Bebian 11 with python 3.9.2.
But when I upgrade to pgadmin4-server 7.6, my uWSGI service pops up with the same error:
`août 31 15:36:10 ERP-NEW uwsgi[4681]: *** Operational MODE: threaded ***
août 31 15:36:14 ERP-NEW uwsgi[4681]: WSGI app 0 (mountpoint='') ready in 4 seconds on interpreter 0x556fdb64a1d0 pid: 4681 (default app)
août 31 15:36:14 ERP-NEW uwsgi[4681]: mounting pgAdmin4:app on /pgadmin
août 31 15:36:15 ERP-NEW uwsgi[4681]: Traceback (most recent call last):
août 31 15:36:15 ERP-NEW uwsgi[4681]:   File "./pgAdmin4.py", line 104, in <module>
août 31 15:36:15 ERP-NEW uwsgi[4681]:     app = create_app()
août 31 15:36:15 ERP-NEW uwsgi[4681]:   File "/usr/pgadmin4/web/pgadmin/__init__.py", line 550, in create_app
août 31 15:36:15 ERP-NEW uwsgi[4681]:     driver.init_app(app)
août 31 15:36:15 ERP-NEW uwsgi[4681]:   File "/usr/pgadmin4/web/pgadmin/utils/driver/__init__.py", line 27, in init_app
août 31 15:36:15 ERP-NEW uwsgi[4681]:     DriverRegistry.load_modules(app)
août 31 15:36:15 ERP-NEW uwsgi[4681]:   File "/usr/pgadmin4/web/pgadmin/utils/driver/registry.py", line 20, in load_modules
août 31 15:36:15 ERP-NEW uwsgi[4681]:     from . import psycopg3 as module
août 31 15:36:15 ERP-NEW uwsgi[4681]:   File "/usr/pgadmin4/web/pgadmin/utils/driver/psycopg3/__init__.py", line 28, in <module>
août 31 15:36:15 ERP-NEW uwsgi[4681]:     from .connection import Connection
août 31 15:36:15 ERP-NEW uwsgi[4681]:   File "/usr/pgadmin4/web/pgadmin/utils/driver/psycopg3/connection.py", line 25, in <module>
août 31 15:36:15 ERP-NEW uwsgi[4681]:     from pgadmin.utils.crypto import decrypt
août 31 15:36:15 ERP-NEW uwsgi[4681]:   File "/usr/pgadmin4/web/pgadmin/utils/crypto.py", line 17, in <module>
août 31 15:36:15 ERP-NEW uwsgi[4681]:     from cryptography.hazmat.primitives.ciphers import Cipher
août 31 15:36:15 ERP-NEW uwsgi[4681]:   File "/usr/pgadmin4/venv/lib/python3.9/site-packages/cryptography/hazmat/primitives/ciphers/__init__.py", line 11, in <module>
août 31 15:36:15 ERP-NEW uwsgi[4681]:     from cryptography.hazmat.primitives.ciphers.base import (
août 31 15:36:15 ERP-NEW uwsgi[4681]:   File "/usr/pgadmin4/venv/lib/python3.9/site-packages/cryptography/hazmat/primitives/ciphers/base.py", line 10, in <module>
août 31 15:36:15 ERP-NEW uwsgi[4681]:     from cryptography.exceptions import (
août 31 15:36:15 ERP-NEW uwsgi[4681]:   File "/usr/pgadmin4/venv/lib/python3.9/site-packages/cryptography/exceptions.py", line 9, in <module>
août 31 15:36:15 ERP-NEW uwsgi[4681]:     from cryptography.hazmat.bindings._rust import exceptions as rust_exceptions
août 31 15:36:15 ERP-NEW uwsgi[4681]: ImportError: PyO3 modules may only be initialized once per interpreter process
août 31 15:36:15 ERP-NEW uwsgi[4681]: unable to load app 1 (mountpoint='/pgadmin') (callable not found or import error)
août 31 15:36:15 ERP-NEW uwsgi[4681]: *** uWSGI is running in multiple interpreter mode ***
`
My PgAdmin server runs through Systemd-uWSGI-Nginx.

**Comment by olivluca** [comment]
> I am not able to reproduce the issue on Debian 11

Mhh, the mail I received from github said  "I **am able** to reproduce the issue" so I disregarded it :cry: 



> Could you please share the exact step to reproduce?

I just installed the package pgadmin4-server, configured it to run under uwsgi (package uwsgi) with the configuration file posted above, then in one nginx server (package nginx-extras) I have this snippet

```
   location = /pgadmin4 {
     return 301 /pgadmin/;
   }

   location ^~ /pgadmin {
     include snippets/lemon.conf;
     include uwsgi_params;
     uwsgi_param REMOTE_USER $lmremote_user;
     uwsgi_pass unix:/run/uwsgi/app/pgadmin4/socket;
   }

```

snippets/lemon.conf is for the lemonldap authentication portal, however I also tested with an hardcoded user and no authentication

```
   location ^~ /pgadmin {
     include uwsgi_params;
     uwsgi_param REMOTE_USER luca;
     uwsgi_pass unix:/run/uwsgi/app/pgadmin4/socket;
   }
```

uwsgi_params is unmodified from the nginx-common package

```
uwsgi_param  QUERY_STRING       $query_string;
uwsgi_param  REQUEST_METHOD     $request_method;
uwsgi_param  CONTENT_TYPE       $content_type;
uwsgi_param  CONTENT_LENGTH     $content_length;

uwsgi_param  REQUEST_URI        $request_uri;
uwsgi_param  PATH_INFO          $document_uri;
uwsgi_param  DOCUMENT_ROOT      $document_root;
uwsgi_param  SERVER_PROTOCOL    $server_protocol;
uwsgi_param  REQUEST_SCHEME     $scheme;
uwsgi_param  HTTPS              $https if_not_empty;

uwsgi_param  REMOTE_ADDR        $remote_addr;
uwsgi_param  REMOTE_PORT        $remote_port;
uwsgi_param  SERVER_PORT        $server_port;
uwsgi_param  SERVER_NAME        $server_name;

```

**Comment by FlossyWeb** [comment]
My Nginx configuration is different
`upstream pgadmin {
  server unix:///run/pgadmin4/pgadmin4.sock;
}
[...]
  # Redirect PgAdmin4
  location ^~ /pgadmin {
    include uwsgi_params;
    uwsgi_pass  pgadmin;
  }
`
Although I don't think it really matters regarding this issue, also I'm not a Python ninja so...

**Comment by yogeshmahajan-1903** [comment]
Issue is due to cryptography version. Downgrading version to 40.0.2, issue is resolved.

**Comment by yogeshmahajan-1903** [comment]
This can be fixed only after -  https://github.com/pyca/cryptography/issues/9016.
Issue is only reproducible if pgAdmin launched using uWSGI service/uwsgi module installed in venv.

**Comment by olivluca** [comment]
For the time being could you do the same as in this commit (mentioned in that bug report)?
https://github.com/zigbeefordomoticz/Domoticz-Zigbee/commit/f8d16ff995207e162cf44812cb49e3b523182822

**Comment by anilsahoo20** [comment]
Tested and verified on Snapshot build: https://www.postgresql.org/ftp/pgadmin/pgadmin4/snapshots/2023-09-29/
Environment: Debian 11(Bullseye) amd64->pgAdmin 4 Server mode

**Comment by olivluca** [comment]
I suppose it's commit f9bbadfc6a49ada0269dd5f36b8c0c84e1c9c766
Thank you for fixing it.

**Comment by olivluca** [comment]
The issue reappeared in debian 11 packages for versions 8.0, 8.1 and 8.2 :disappointed: 
Back to 7.8 (after manually correcting issue #6874)

**Comment by olivluca** [comment]
Same error in 8.12

**Comment by yogeshmahajan-1903** [comment]
@olivluca 
Can you please share the latest pgadmin logs? Also pip freeze of virtual environment being used in uWSGI service?

**Comment by olivluca** [comment]
The log below is the same as the one in the initial report, the cause is in https://github.com/pgadmin-org/pgadmin4/issues/6718#issuecomment-1729227604, the solution (I think) is in https://github.com/pgadmin-org/pgadmin4/commit/f9bbadfc6a49ada0269dd5f36b8c0c84e1c9c766 (while now it is pinned at 43.0), I don't know how to get the "pip freeze".

```
Tue Oct 22 09:04:40 2024 - *** Starting uWSGI 2.0.19.1-debian (64bit) on [Tue Oct 22 09:04:40 2024] ***
Tue Oct 22 09:04:40 2024 - compiled with version: 10.2.1 20210110 on 11 June 2021 09:08:33
Tue Oct 22 09:04:40 2024 - os: Linux-5.10.0-33-amd64 #1 SMP Debian 5.10.226-1 (2024-10-03)
Tue Oct 22 09:04:40 2024 - nodename: web
Tue Oct 22 09:04:40 2024 - machine: x86_64
Tue Oct 22 09:04:40 2024 - clock source: unix
Tue Oct 22 09:04:40 2024 - pcre jit disabled
Tue Oct 22 09:04:40 2024 - detected number of CPU cores: 2
Tue Oct 22 09:04:40 2024 - current working directory: /
Tue Oct 22 09:04:40 2024 - writing pidfile to /run/uwsgi/app/pgadmin4/pid
Tue Oct 22 09:04:40 2024 - detected binary path: /usr/bin/uwsgi-core
Tue Oct 22 09:04:40 2024 - chdir() to /usr/pgadmin4/web/
Tue Oct 22 09:04:40 2024 - your processes number limit is 15538
Tue Oct 22 09:04:40 2024 - your memory page size is 4096 bytes
Tue Oct 22 09:04:40 2024 - detected max file descriptor number: 1024
Tue Oct 22 09:04:40 2024 - lock engine: pthread robust mutexes
Tue Oct 22 09:04:40 2024 - thunder lock: disabled (you can enable it with --thunder-lock)
Tue Oct 22 09:04:40 2024 - uwsgi socket 0 bound to UNIX address /run/uwsgi/app/pgadmin4/socket fd 3
Tue Oct 22 09:04:40 2024 - setgid() to 33
Tue Oct 22 09:04:40 2024 - setuid() to 114
Tue Oct 22 09:04:40 2024 - Python version: 3.9.2 (default, Feb 28 2021, 17:03:44)  [GCC 10.2.1 20210110]
Tue Oct 22 09:04:40 2024 - PEP 405 virtualenv detected: /usr/pgadmin4/venv
Tue Oct 22 09:04:40 2024 - Set PythonHome to /usr/pgadmin4/venv
Tue Oct 22 09:04:40 2024 - Python main interpreter initialized at 0x55f4ade6b430
Tue Oct 22 09:04:40 2024 - python threads support enabled
Tue Oct 22 09:04:40 2024 - your server socket listen backlog is limited to 100 connections
Tue Oct 22 09:04:40 2024 - your mercy for graceful operations on workers is 60 seconds
Tue Oct 22 09:04:40 2024 - mapped 648528 bytes (633 KB) for 25 cores
Tue Oct 22 09:04:40 2024 - *** Operational MODE: threaded ***
Tue Oct 22 09:05:37 2024 - WSGI app 0 (mountpoint='') ready in 57 seconds on interpreter 0x55f4ade6b430 pid: 1400 (default app)
Tue Oct 22 09:05:37 2024 - mounting pgAdmin4:app on /pgadmin
Traceback (most recent call last):
  File "./pgAdmin4.py", line 42, in <module>
    from eventlet import hubs
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/__init__.py", line 11, in <module>
    from eventlet import convenience
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/convenience.py", line 7, in <module>
    from eventlet.green import socket
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/green/socket.py", line 21, in <module>
    from eventlet.support import greendns
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/support/greendns.py", line 79, in <module>
    setattr(dns, pkg, import_patched('dns.' + pkg))
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/support/greendns.py", line 61, in import_patched
    return patcher.import_patched(module_name, **modules)
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/patcher.py", line 132, in import_patched
    return inject(
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/eventlet/patcher.py", line 109, in inject
    module = __import__(module_name, {}, {}, module_name.split('.')[:-1])
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/dns/dnssec.py", line 1174, in <module>
    from cryptography.exceptions import InvalidSignature
  File "/usr/pgadmin4/venv/lib/python3.9/site-packages/cryptography/exceptions.py", line 9, in <module>
    from cryptography.hazmat.bindings._rust import exceptions as rust_exceptions
ImportError: PyO3 modules do not yet support subinterpreters, see https://github.com/PyO3/pyo3/issues/576
Tue Oct 22 09:05:37 2024 - unable to load app 1 (mountpoint='/pgadmin') (callable not found or import error)
Tue Oct 22 09:05:37 2024 - *** uWSGI is running in multiple interpreter mode ***
Tue Oct 22 09:05:37 2024 - spawned uWSGI master process (pid: 1400)
Tue Oct 22 09:05:37 2024 - spawned uWSGI worker 1 (pid: 1914, cores: 25)

```

**Comment by olivluca** [comment]
OK, to "fix" it downgrading cryptography to 40.0.2 wasn't enough, I also had to downgrade bcrypt to 4.0.1

**Comment by GreenReaper** [comment]
I appear to be getting the same as @olivluca, also with Bullseye. I have pgadmin4-server 9.8 installed from the PostgreSQL APT repository.

Looking in `/usr/pgadmin4/venv/lib/python3.11/site-packages/` I see `cryptography-45.0.7` and `bcrypt-4.3.0` mentioned.

It wasn't clear to me how to downgrade these to the versions mentioned above but I eventually figured it out (at least, this got me to the login screen and then in):
```
chown -R pgadmin4:pgadmin4 /usr/pgadmin4
su - pgadmin4
cd /usr/pgadmin4
source venv/bin/activate
pip3 install cryptography==40.0.2
pip3 install bcrypt==4.0.1
[restart uwsgi]
```

Initially I didn't have /usr/pgadmin4 owned by pgadmin4 which gave
```
Defaulting to user installation because normal site-packages is not writeable
...
ERROR: Will not install to the user site because it will lack sys.path precedence to cryptography in /usr/pgadmin4/venv/lib/python3.11/site-packages
```
I am of course a little concerned that this may cause further errors, or a security issue - is there some other solution for this, or some way to keep it working beyond holding on this tweaked version of 9.8? I appreciate the normal server is Apache but ideally we wouldn't run a whole other web server. Or perhaps "the DEB and RPM package maintainers are informed" bit from Requirements.txt needs to be done?

(I think the underlying issue is PyO3/pyo3/issues/3451 which seems to be making progress, but will need to filter through to pycryptography as well.)
